local_storage.go

```diff
@@ -33,66 +33,141 @@ func newItemNotFoundError(format string, args ...interface{}) ItemNotFoundError
 	}
 }
 
-// SetItem saves the given item in localStorage under the given key.
-func SetItem(key, item string) error {
+func init() {
+	DetectStorage()
+}
+
+// DetectStorage detects and (re)initializes the localStorage.
+func DetectStorage() (ok bool) {
+	defer func() {
+		if r := recover(); r != nil {
+			localStorage = nil
+			ok = false
+		}
+	}()
+
+	localStorage = js.Global.Get("localStorage")
+
 	if localStorage == js.Undefined {
+		localStorage = nil
+	}
+	if localStorage == nil {
+		return false
+	}
+
+	// Cf. https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
+	// https://gist.github.com/paulirish/5558557
+	x := "__storage_test__"
+	localStorage.Set(x, x)
+	obj := localStorage.Get(x)
+	if obj == js.Undefined || obj == nil {
+		localStorage = nil
+		return false
+	}
+	localStorage.Call("removeItem", x)
+	return true
+}
+
+// SetItem saves the given item in localStorage under the given key.
+func SetItem(key, item string) (err error) {
+	if localStorage == nil || localStorage == js.Undefined {
 		return ErrLocalStorageNotSupported
 	}
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
+		}
+	}()
+
 	localStorage.Call("setItem", key, item)
 	return nil
 }
 
 // GetItem finds and returns the item identified by key. If there is no item in
 // localStorage with the given key, GetItem will return an ItemNotFoundError.
-func GetItem(key string) (string, error) {
-	if localStorage == js.Undefined {
+func GetItem(key string) (s string, err error) {
+	if localStorage == nil || localStorage == js.Undefined {
 		return "", ErrLocalStorageNotSupported
 	}
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
```

Comment:
1. Correct me if I'm wrong, but I don't think we can be certain that whatever was returned by `recover` has type `error`. You can technically pass anything to `panic`, and given the complexity of what GopherJS is doing, I wouldn't be surprised if we end up receiving a `string` or something else here sometimes. I'd like to check if `r` is type `error` and handle it gracefully if it is not.
2. This is nit-picky, but I usually like to return the zero value for the first return value whenever there is an error. This helps keep things tidy and prevents users from using a value which may be invalid or inconsistent in the case of an error. Can we add `s = ""` here before returning?

So the final version should look like:

```
if r := recover(); r != nil {
    var ok bool
    err, ok = r.(error)
    if !ok {
        err = fmt.Errorf("go-humble/router recovered from panic with non-error type: %v", r)
    }
    s = ""
}
```

This same feedback applies to other places in the code where we recover from panics.